File: backend/src/services/nft_avatar_service.py
"""
NFT Avatar Service

This module provides services for interacting with the NFTAvatarManager
smart contract and managing NFT avatar data.

Author: Manus AI
Date: 2025-11-05
"""

import os
from web3 import Web3
from datetime import datetime
from typing import Dict, List, Optional, Tuple
from ..models.subscription import NFTAvatar, NFTStandard, db


class NFTAvatarService:
    """
    Service for managing NFT avatars via smart contract
    """
    
    def __init__(self):
        """Initialize Web3 connection and contract"""
        # Connect to Web3 provider
        provider_url = os.getenv('WEB3_PROVIDER_URL', 'https://sepolia.infura.io/v3/YOUR_INFURA_KEY')
        self.w3 = Web3(Web3.HTTPProvider(provider_url))
        
        # Contract address
        self.nft_avatar_contract_address = os.getenv(
            'CONTRACT_NFT_AVATAR_MANAGER',
            '0xF91E0E6afF5A93831F67838539245a44Ca384187'
        )
        
        # Load contract ABI
        self.nft_avatar_abi = self._load_nft_avatar_abi()
        
        # Initialize contract
        self.nft_avatar_contract = self.w3.eth.contract(
            address=self.nft_avatar_contract_address,
            abi=self.nft_avatar_abi
        )
    
    def _load_nft_avatar_abi(self) -> List[Dict]:
        """
        Load NFTAvatarManager contract ABI
        
        Returns:
            Contract ABI as list of dictionaries
        """
        # In production, load from file or environment
        # For now, return minimal ABI for key functions
        return [
            {
                "inputs": [{"internalType": "address", "name": "user", "type": "address"}],
                "name": "getUserAvatar",
                "outputs": [
                    {"internalType": "address", "name": "contractAddress", "type": "address"},
                    {"internalType": "uint256", "name": "tokenId", "type": "uint256"},
                    {"internalType": "uint8", "name": "standard", "type": "uint8"},
                    {"internalType": "uint256", "name": "setAt", "type": "uint256"}
                ],
                "stateMutability": "view",
                "type": "function"
            },
            {
                "inputs": [{"internalType": "address", "name": "user", "type": "address"}],
                "name": "verifyAvatarOwnership",
                "outputs": [{"internalType": "bool", "name": "", "type": "bool"}],
                "stateMutability": "view",
                "type": "function"
            },
            {
                "inputs": [{"internalType": "address", "name": "user", "type": "address"}],
                "name": "getUserAvatarHistory",
                "outputs": [
                    {
                        "components": [
                            {"internalType": "address", "name": "contractAddress", "type": "address"},
                            {"internalType": "uint256", "name": "tokenId", "type": "uint256"},
                            {"internalType": "uint8", "name": "standard", "type": "uint8"},
                            {"internalType": "uint256", "name": "setAt", "type": "uint256"}
                        ],
                        "internalType": "struct NFTAvatarManager.NFTAvatar[]",
                        "name": "",
                        "type": "tuple[]"
                    }
                ],
                "stateMutability": "view",
                "type": "function"
            }
        ]
    
    def get_user_avatar(self, user_address: str) -> Optional[Dict]:
        """
        Get user's current NFT avatar from blockchain
        
        Args:
            user_address: User's wallet address
            
        Returns:
            Avatar data dictionary or None if no avatar set
        """
        try:
            # Call smart contract
            result = self.nft_avatar_contract.functions.getUserAvatar(
                Web3.to_checksum_address(user_address)
            ).call()
            
            contract_address, token_id, standard, set_at = result
            
            # Return None if no avatar set
            if contract_address == '0x0000000000000000000000000000000000000000':
                return None
            
            # Map standard enum
            standard_map = {0: 'ERC721', 1: 'ERC1155'}
            
            return {
                'contractAddress': contract_address,
                'tokenId': str(token_id),
                'standard': standard_map.get(standard, 'ERC721'),
                'setAt': datetime.fromtimestamp(set_at)
            }
            
        except Exception as e:
            print(f"Error getting user avatar: {e}")
            return None
    
    def verify_avatar_ownership(self, user_address: str) -> bool:
        """
        Verify if user still owns their NFT avatar
        
        Args:
            user_address: User's wallet address
            
        Returns:
            True if user owns the avatar, False otherwise
        """
        try:
            return self.nft_avatar_contract.functions.verifyAvatarOwnership(
                Web3.to_checksum_address(user_address)
            ).call()
        except Exception as e:
            print(f"Error verifying avatar ownership: {e}")
            return False
    
    def get_user_avatar_history(self, user_address: str) -> List[Dict]:
        """
        Get user's avatar history from blockchain
        
        Args:
            user_address: User's wallet address
            
        Returns:
            List of avatar data dictionaries
        """
        try:
            # Call smart contract
            results = self.nft_avatar_contract.functions.getUserAvatarHistory(
                Web3.to_checksum_address(user_address)
            ).call()
            
            # Map standard enum
            standard_map = {0: 'ERC721', 1: 'ERC1155'}
            
            avatars = []
            for contract_address, token_id, standard, set_at in results:
                avatars.append({
                    'contractAddress': contract_address,
                    'tokenId': str(token_id),
                    'standard': standard_map.get(standard, 'ERC721'),
                    'setAt': datetime.fromtimestamp(set_at)
                })
            
            return avatars
            
        except Exception as e:
            print(f"Error getting avatar history: {e}")
            return []
    
    def verify_transaction(self, tx_hash: str) -> Tuple[bool, Optional[Dict]]:
        """
        Verify a transaction on the blockchain
        
        Args:
            tx_hash: Transaction hash
            
        Returns:
            Tuple of (success, receipt_data)
        """
        try:
            # Get transaction receipt
            receipt = self.w3.eth.get_transaction_receipt(tx_hash)
            
            # Check if transaction was successful
            if receipt['status'] != 1:
                return False, None
            
            # Parse receipt data
            receipt_data = {
                'transactionHash': receipt['transactionHash'].hex(),
                'blockNumber': receipt['blockNumber'],
                'from': receipt['from'],
                'to': receipt['to'],
                'gasUsed': receipt['gasUsed'],
                'status': receipt['status']
            }
            
            return True, receipt_data
            
        except Exception as e:
            print(f"Error verifying transaction: {e}")
            return False, None
    
    def sync_avatar_to_db(self, user_address: str, tx_hash: str) -> Optional[NFTAvatar]:
        """
        Sync avatar data from blockchain to database
        
        Args:
            user_address: User's wallet address
            tx_hash: Transaction hash
            
        Returns:
            NFTAvatar object or None if failed
        """
        try:
            # Get avatar from blockchain
            avatar_data = self.get_user_avatar(user_address)
            if not avatar_data:
                return None
            
            # Mark all previous avatars as not current
            NFTAvatar.query.filter_by(
                user_address=user_address.lower(),
                is_current=True
            ).update({'is_current': False})
            
            # Check if this avatar already exists
            existing_avatar = NFTAvatar.query.filter_by(
                user_address=user_address.lower(),
                contract_address=avatar_data['contractAddress'].lower(),
                token_id=avatar_data['tokenId']
            ).first()
            
            if existing_avatar:
                # Update existing avatar
                existing_avatar.is_current = True
                existing_avatar.is_valid = True
                existing_avatar.set_at = avatar_data['setAt']
                existing_avatar.transaction_hash = tx_hash
                existing_avatar.updated_at = datetime.utcnow()
                db.session.commit()
                return existing_avatar
            else:
                # Create new avatar
                new_avatar = NFTAvatar(
                    user_address=user_address.lower(),
                    contract_address=avatar_data['contractAddress'].lower(),
                    token_id=avatar_data['tokenId'],
                    standard=NFTStandard[avatar_data['standard']],
                    is_current=True,
                    is_valid=True,
                    transaction_hash=tx_hash,
                    set_at=avatar_data['setAt']
                )
                db.session.add(new_avatar)
                db.session.commit()
                return new_avatar
                
        except Exception as e:
            print(f"Error syncing avatar to DB: {e}")
            db.session.rollback()
            return None
    
    def get_user_avatar_from_db(self, user_address: str) -> Optional[Dict]:
        """
        Get user's current avatar from database
        
        Args:
            user_address: User's wallet address
            
        Returns:
            Avatar dictionary or None
        """
        try:
            avatar = NFTAvatar.query.filter_by(
                user_address=user_address.lower(),
                is_current=True
            ).first()
            
            if avatar:
                # Verify ownership on blockchain
                is_valid = self.verify_avatar_ownership(user_address)
                if avatar.is_valid != is_valid:
                    avatar.is_valid = is_valid
                    db.session.commit()
                
                return avatar.to_dict()
            
            return None
            
        except Exception as e:
            print(f"Error getting avatar from DB: {e}")
            return None
    
    def get_user_avatar_history_from_db(self, user_address: str) -> List[Dict]:
        """
        Get user's avatar history from database
        
        Args:
            user_address: User's wallet address
            
        Returns:
            List of avatar dictionaries
        """
        try:
            avatars = NFTAvatar.query.filter_by(
                user_address=user_address.lower()
            ).order_by(NFTAvatar.set_at.desc()).all()
            
            return [avatar.to_dict() for avatar in avatars]
            
        except Exception as e:
            print(f"Error getting avatar history from DB: {e}")
            return []


# Singleton instance
nft_avatar_service = NFTAvatarService()

File: dchat-main/frontend/src/services/NFTAvatarService.js
/**
 * NFT Avatar Service
 * 
 * This service handles interactions with the NFTAvatarManager smart contract
 * and provides methods for setting and managing NFT avatars.
 * 
 * Features:
 * - Set ERC-721 or ERC-1155 NFT as avatar
 * - Remove NFT avatar
 * - Get avatar history
 * - Verify NFT ownership
 * 
 * @author Manus AI
 * @date 2025-11-05
 */

import { ethers } from 'ethers'

// Contract address (Sepolia testnet)
const NFT_AVATAR_MANAGER_ADDRESS = '0xF91E0E6afF5A93831F67838539245a44Ca384187'

// NFT standards
export const NFT_STANDARDS = {
  ERC721: 0,
  ERC1155: 1
}

// Minimal ABI for NFTAvatarManager contract
const NFT_AVATAR_MANAGER_ABI = [
  // Set ERC-721 avatar
  {
    "inputs": [
      {"internalType": "address", "name": "nftContract", "type": "address"},
      {"internalType": "uint256", "name": "tokenId", "type": "uint256"}
    ],
    "name": "setAvatarERC721",
    "outputs": [],
    "stateMutability": "nonpayable",
    "type": "function"
  },
  // Set ERC-1155 avatar
  {
    "inputs": [
      {"internalType": "address", "name": "nftContract", "type": "address"},
      {"internalType": "uint256", "name": "tokenId", "type": "uint256"}
    ],
    "name": "setAvatarERC1155",
    "outputs": [],
    "stateMutability": "nonpayable",
    "type": "function"
  },
  // Remove avatar
  {
    "inputs": [],
    "name": "removeAvatar",
    "outputs": [],
    "stateMutability": "nonpayable",
    "type": "function"
  },
  // Get user avatar
  {
    "inputs": [{"internalType": "address", "name": "user", "type": "address"}],
    "name": "getUserAvatar",
    "outputs": [
      {"internalType": "address", "name": "contractAddress", "type": "address"},
      {"internalType": "uint256", "name": "tokenId", "type": "uint256"},
      {"internalType": "uint8", "name": "standard", "type": "uint8"},
      {"internalType": "uint256", "name": "setAt", "type": "uint256"}
    ],
    "stateMutability": "view",
    "type": "function"
  },
  // Verify avatar ownership
  {
    "inputs": [{"internalType": "address", "name": "user", "type": "address"}],
    "name": "verifyAvatarOwnership",
    "outputs": [{"internalType": "bool", "name": "", "type": "bool"}],
    "stateMutability": "view",
    "type": "function"
  },
  // Get avatar history
  {
    "inputs": [{"internalType": "address", "name": "user", "type": "address"}],
    "name": "getUserAvatarHistory",
    "outputs": [
      {
        "components": [
          {"internalType": "address", "name": "contractAddress", "type": "address"},
          {"internalType": "uint256", "name": "tokenId", "type": "uint256"},
          {"internalType": "uint8", "name": "standard", "type": "uint8"},
          {"internalType": "uint256", "name": "setAt", "type": "uint256"}
        ],
        "internalType": "struct NFTAvatarManager.NFTAvatar[]",
        "name": "",
        "type": "tuple[]"
      }
    ],
    "stateMutability": "view",
    "type": "function"
  }
]

// ERC-721 ABI (for checking ownership)
const ERC721_ABI = [
  {
    "inputs": [{"internalType": "uint256", "name": "tokenId", "type": "uint256"}],
    "name": "ownerOf",
    "outputs": [{"internalType": "address", "name": "", "type": "address"}],
    "stateMutability": "view",
    "type": "function"
  },
  {
    "inputs": [{"internalType": "address", "name": "owner", "type": "address"}],
    "name": "balanceOf",
    "outputs": [{"internalType": "uint256", "name": "", "type": "uint256"}],
    "stateMutability": "view",
    "type": "function"
  }
]

// ERC-1155 ABI (for checking ownership)
const ERC1155_ABI = [
  {
    "inputs": [
      {"internalType": "address", "name": "account", "type": "address"},
      {"internalType": "uint256", "name": "id", "type": "uint256"}
    ],
    "name": "balanceOf",
    "outputs": [{"internalType": "uint256", "name": "", "type": "uint256"}],
    "stateMutability": "view",
    "type": "function"
  }
]

class NFTAvatarService {
  constructor() {
    this.provider = null
    this.signer = null
    this.nftAvatarContract = null
    this.backendUrl = process.env.REACT_APP_API_URL || 'http://localhost:5000'
  }

  /**
   * Initialize the service with Web3 provider
   * @param {Object} provider - Ethereum provider (e.g., from MetaMask)
   */
  async initialize(provider) {
    try {
      this.provider = new ethers.providers.Web3Provider(provider)
      this.signer = this.provider.getSigner()
      
      // Initialize NFT avatar contract
      this.nftAvatarContract = new ethers.Contract(
        NFT_AVATAR_MANAGER_ADDRESS,
        NFT_AVATAR_MANAGER_ABI,
        this.signer
      )
      
      console.log('✅ NFTAvatarService initialized')
      return true
    } catch (error) {
      console.error('❌ Failed to initialize NFTAvatarService:', error)
      return false
    }
  }

  /**
   * Get user's NFT avatar
   * @param {string} userAddress - User's wallet address
   * @returns {Object|null} Avatar information or null if no avatar set
   */
  async getUserAvatar(userAddress) {
    try {
      const avatar = await this.nftAvatarContract.getUserAvatar(userAddress)
      
      // Check if avatar is set (contract address is not zero)
      if (avatar.contractAddress === '0x0000000000000000000000000000000000000000') {
        return null
      }
      
      return {
        contractAddress: avatar.contractAddress,
        tokenId: avatar.tokenId.toString(),
        standard: avatar.standard === NFT_STANDARDS.ERC721 ? 'ERC721' : 'ERC1155',
        setAt: new Date(avatar.setAt.toNumber() * 1000)
      }
    } catch (error) {
      console.error('Error getting user avatar:', error)
      return null
    }
  }

  /**
   * Verify if user still owns their NFT avatar
   * @param {string} userAddress - User's wallet address
   * @returns {boolean} True if user owns the avatar
   */
  async verifyAvatarOwnership(userAddress) {
    try {
      return await this.nftAvatarContract.verifyAvatarOwnership(userAddress)
    } catch (error) {
      console.error('Error verifying avatar ownership:', error)
      return false
    }
  }

  /**
   * Get user's avatar history
   * @param {string} userAddress - User's wallet address
   * @returns {Array} Array of avatar objects
   */
  async getUserAvatarHistory(userAddress) {
    try {
      const history = await this.nftAvatarContract.getUserAvatarHistory(userAddress)
      
      return history.map(avatar => ({
        contractAddress: avatar.contractAddress,
        tokenId: avatar.tokenId.toString(),
        standard: avatar.standard === NFT_STANDARDS.ERC721 ? 'ERC721' : 'ERC1155',
        setAt: new Date(avatar.setAt.toNumber() * 1000)
      }))
    } catch (error) {
      console.error('Error getting avatar history:', error)
      return []
    }
  }

  /**
   * Check if user owns an NFT
   * @param {string} nftContract - NFT contract address
   * @param {string} tokenId - Token ID
   * @param {string} standard - NFT standard ('ERC721' or 'ERC1155')
   * @param {string} userAddress - User's wallet address
   * @returns {boolean} True if user owns the NFT
   */
  async checkNFTOwnership(nftContract, tokenId, standard, userAddress) {
    try {
      if (standard === 'ERC721') {
        const contract = new ethers.Contract(nftContract, ERC721_ABI, this.provider)
        const owner = await contract.ownerOf(tokenId)
        return owner.toLowerCase() === userAddress.toLowerCase()
      } else if (standard === 'ERC1155') {
        const contract = new ethers.Contract(nftContract, ERC1155_ABI, this.provider)
        const balance = await contract.balanceOf(userAddress, tokenId)
        return balance.gt(0)
      }
      return false
    } catch (error) {
      console.error('Error checking NFT ownership:', error)
      return false
    }
  }

  /**
   * Set NFT as avatar
   * @param {string} nftContract - NFT contract address
   * @param {string} tokenId - Token ID
   * @param {string} standard - NFT standard ('ERC721' or 'ERC1155')
   * @param {Function} onProgress - Progress callback
   * @returns {Object} Transaction receipt
   */
  async setNFTAvatar(nftContract, tokenId, standard, onProgress = null) {
    try {
      const userAddress = await this.signer.getAddress()
      
      // Step 1: Verify ownership
      onProgress?.({ step: 1, message: 'Verifying NFT ownership...' })
      const ownsNFT = await this.checkNFTOwnership(nftContract, tokenId, standard, userAddress)
      if (!ownsNFT) {
        throw new Error('You do not own this NFT')
      }
      
      // Step 2: Set avatar on blockchain
      onProgress?.({ step: 2, message: 'Setting NFT avatar...' })
      let tx
      if (standard === 'ERC721') {
        tx = await this.nftAvatarContract.setAvatarERC721(nftContract, tokenId)
      } else if (standard === 'ERC1155') {
        tx = await this.nftAvatarContract.setAvatarERC1155(nftContract, tokenId)
      } else {
        throw new Error('Invalid NFT standard')
      }
      
      // Step 3: Wait for confirmation
      onProgress?.({ step: 3, message: 'Waiting for confirmation...' })
      const receipt = await tx.wait()
      
      // Step 4: Sync with backend
      onProgress?.({ step: 4, message: 'Syncing with backend...' })
      await this.syncAvatarToBackend(receipt.transactionHash, nftContract, tokenId, standard)
      
      onProgress?.({ step: 5, message: 'Avatar set successfully!' })
      
      return {
        success: true,
        transactionHash: receipt.transactionHash
      }
    } catch (error) {
      console.error('Error setting NFT avatar:', error)
      throw error
    }
  }

  /**
   * Remove NFT avatar
   * @returns {Object} Transaction receipt
   */
  async removeNFTAvatar() {
    try {
      const tx = await this.nftAvatarContract.removeAvatar()
      const receipt = await tx.wait()
      
      return {
        success: true,
        transactionHash: receipt.transactionHash
      }
    } catch (error) {
      console.error('Error removing NFT avatar:', error)
      throw error
    }
  }

  /**
   * Sync avatar to backend
   * @param {string} transactionHash - Transaction hash
   * @param {string} nftContract - NFT contract address
   * @param {string} tokenId - Token ID
   * @param {string} standard - NFT standard
   */
  async syncAvatarToBackend(transactionHash, nftContract, tokenId, standard) {
    try {
      const response = await fetch(`${this.backendUrl}/api/avatars/nft/set`, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
          'Authorization': `Bearer ${localStorage.getItem('jwt_token')}`,
          'X-User-Address': await this.signer.getAddress()
        },
        body: JSON.stringify({
          nftContract,
          tokenId,
          standard,
          transactionHash
        })
      })
      
      if (!response.ok) {
        throw new Error('Failed to sync avatar to backend')
      }
      
      const data = await response.json()
      console.log('✅ Avatar synced to backend:', data)
      return data
    } catch (error) {
      console.error('Error syncing avatar to backend:', error)
      // Don't throw error - avatar is still valid on blockchain
    }
  }

  /**
   * Get avatar from backend
   * @param {string} userAddress - User's wallet address
   * @returns {Object|null} Avatar information
   */
  async getAvatarFromBackend(userAddress) {
    try {
      const response = await fetch(`${this.backendUrl}/api/avatars/nft/${userAddress}`)
      
      if (!response.ok) {
        return null
      }
      
      const data = await response.json()
      return data.avatar
    } catch (error) {
      console.error('Error getting avatar from backend:', error)
      return null
    }
  }

  /**
   * Get user's NFTs from wallet
   * This is a placeholder - in production, you would use services like
   * Alchemy, Moralis, or OpenSea API to fetch user's NFTs
   * 
   * @param {string} userAddress - User's wallet address
   * @returns {Array} Array of NFT objects
   */
  async getUserNFTs(userAddress) {
    try {
      // TODO: Implement NFT fetching using Alchemy/Moralis API
      // For now, return empty array
      console.log('TODO: Implement getUserNFTs using Alchemy/Moralis API')
      return []
    } catch (error) {
      console.error('Error getting user NFTs:', error)
      return []
    }
  }

  /**
   * Get NFT metadata
   * @param {string} nftContract - NFT contract address
   * @param {string} tokenId - Token ID
   * @returns {Object} NFT metadata
   */
  async getNFTMetadata(nftContract, tokenId) {
    try {
      // TODO: Implement metadata fetching
      // For now, return placeholder
      return {
        name: `NFT #${tokenId}`,
        description: 'NFT Avatar',
        image: null
      }
    } catch (error) {
      console.error('Error getting NFT metadata:', error)
      return null
    }
  }
}

// Export singleton instance
export const nftAvatarService = new NFTAvatarService()
export default nftAvatarService
